### PR TITLE
Generalise compiled decode segments to Qwen 3 Next

### PR DESCRIPTION
**Pull request by aleroot**
## Proposed changes

Qwen 3 Next previously executed each decoder layer separately during single-token generation, which adds dispatch overhead and slows token generation. Qwen 3.5 already reduced this overhead using compiled decode segments, but that implementation was model-specific.

This PR:

- Extracts the decode-segment schedule and thread-safe compiled-function cache into `MLXLMCommon`.
- Updates Qwen 3.5 to use the shared implementation without changing its behaviour.
- Adds compiled single-token decode segments to Qwen 3 Next.
- Keeps changing KV-cache operations outside compiled graphs.
- Supports fp16 activations, including models with 4-bit quantised weights.
- Falls back to the existing implementation for bfloat16, float32, quantised KV caches, and TurboQuant KV caches where exact equivalence has not been established.

Tests compare generated logits and every cache tensor byte-for-byte between the compiled and existing paths. They also cover quantized weights, unsupported-cache fallback, segment scheduling, and model deallocation.

On a warmed synthetic 16-layer Qwen 3 Next model with 4-bit weights, interleaved measurements showed approximately 35–38% lower median decode latency, corresponding to 1.55–1.60× throughput. Compilation time was excluded.

## Checklist

- [x] I have read the [[CONTRIBUTING](https://github.com/ml-explore/mlx-swift-lm/blob/main/CONTRIBUTING.md)](https://github.com/ml-explore/mlx-swift-lm/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by ngutech21** [comment]
@aleroot Would it make sense to use the new shared `CompiledDecodeSegment` and `CompiledDecodeSegmentCache` infrastructure for the Qwen 3.6 MoE VLM path in `MLXVLM/Models/Qwen35.swift` as well?
That path still lacks the compiled single-token decode segments and GDN decode-convolution folding introduced by #467 and #468. 
Would extending this PR to cover that path be in scope, or would you prefer a separate follow-up PR?

**Comment by aleroot** [comment]
@ngutech21 I think you are right, the VLM path should reuse the new shared infrastructure. I would go with a new PR because this is much more than swapping in CompiledDecodeSegmentCache. The VLM implementation still needs: the traced layer bodies, GDN state plumbing and decodeConv folding from #467/#468, plus VLM-specific handling for M-RoPE/positionIds around the KV-cache update.
Once #569 lands, the follow-up should build on the shared infrastructure rather than duplicating the scheduler. I would  include bitwise logits/cache parity for normal text decode and image-prefill → text-decode continuation, fallback coverage for unsupported cache types.

Is it something you can take care of?

**Comment by ngutech21** [comment]
Sure I can tackle this once #569 lands

File: Libraries/MLXLLM/Models/Qwen3Next.swift [hunk 1]
```diff
@@ -466,6 +590,133 @@ public class Qwen3NextModelInner: Module {
...
+        for (segmentIndex, segment) in decodeSegments.enumerated() {
+            var arguments = [carry] + pendingAttention
+            for layerIndex in segment.linearLayers {
+                let mambaCache = mambaCaches[layerIndex]!
+                arguments.append(mambaCache[0]!)
+                arguments.append(mambaCache[1]!)
+            }
+
+            let outputs = compiledSegments.call(
+                at: segmentIndex, arguments: arguments
```
Comment (davidkoski) [inline Libraries/MLXLLM/Models/Qwen3Next.swift:672]: I was recently working on some `compile` code in mlx-swift and came across something that might matter here: this is (I think) capturing all MLXArrays that are not passed as arguments.  Specifically the layers referenced from `segmentBody`.

This is fine for inference, but if combined with LoRA it likely doesn't work as expected.

I don't think this is new to this PR -- I suspect we have other code in here that does the same thing.

Comment (davidkoski) [reply Libraries/MLXLLM/Models/Qwen3Next.swift:672]: Filed https://github.com/ml-explore/mlx-swift-lm/issues/586 to track this.

Comment (davidkoski) [reply Libraries/MLXLLM/Models/Qwen3Next.swift:672]: I think we are ok to take this, but see that issue.  I have a repro in mlx-swift and it is a potential problem.  It looks like the only instances of this are here and in Qwen35.swift via #469 

Probably worth followup.

**Review by davidkoski — CHANGES_REQUESTED** [review]
Change looks good but see the comment on compile -- I think this is ok to merge as-is and deal with it in a followup, but want to get your take on it.

Thanks!

File: Libraries/MLXLLM/Models/Qwen3Next.swift (hunk 1 above)
Comment (aleroot) [reply Libraries/MLXLLM/Models/Qwen3Next.swift:672]: Sure I will clean this up tomorrow.

**Comment by aleroot** [comment]
> Change looks good but see the comment on compile -- I think this is ok to merge as-is and deal with it in a followup, but want to get your take on it.
> 
> Thanks!

Yes, this is worth processing as a separate issue, tomorrow I think I will have the time to tackle this properly. Thanks.

**Review by davidkoski — APPROVED** [review]
Changes look good.  Compile item to be addressed later -- does not affect inference path and is not new.

**Comment by aleroot** [comment]
Added #589 as follow up to fix the related bug #586
